src/lib.rs

```diff
@@ -115,10 +120,131 @@ impl ZcashScript for Cxx {
+/// differ and always returns the C++ result.
+fn check_legacy_sigop_count_script<T: ZcashScript, U: ZcashScript>(
+    script: &[u8],
+) -> (Result<u32, Error>, Result<u32, Error>) {
+    (
+        T::legacy_sigop_count_script(script),
+        U::legacy_sigop_count_script(script),
+    )
+}
+
+/// Runs two implementations of `ZcashScript::verify_callback` with the same arguments and returns
+/// both results. This is more useful for testing than the impl that logs a warning if the results
+/// differ and always returns the `T` result.
+pub fn check_verify_callback<T: ZcashScript, U: ZcashScript>(
+    sighash: SighashCalculator,
+    lock_time: i64,
+    is_final: bool,
+    script_pub_key: &[u8],
+    script_sig: &[u8],
+    flags: VerificationFlags,
+) -> (Result<(), Error>, Result<(), Error>) {
+    (
+        T::verify_callback(
+            sighash,
+            lock_time,
+            is_final,
+            script_pub_key,
+            script_sig,
+            flags,
+        ),
+        U::verify_callback(
+            sighash,
+            lock_time,
+            is_final,
+            script_pub_key,
+            script_sig,
+            flags,
+        ),
+    )
+}
+
+/// Convert errors that don’t exist in the C++ code into the cases that do.
+pub fn normalize_error(err: Error) -> Error {
+    match err {
+        Error::Ok(Some(_)) => Error::Ok(None),
+        _ => err,
+    }
+}
+
+/// A tag to indicate that both the C++ and Rust implementations of zcash_script should be used,
+/// with their results compared.
+pub enum CxxRustComparisonInterpreter {}
+
+/// This implementation is functionally equivalent to the `T` impl, but it also runs a second (`U`)
+/// impl and logs a warning if they disagree.
+impl ZcashScript for CxxRustComparisonInterpreter {
+    fn legacy_sigop_count_script(script: &[u8]) -> Result<u32, Error> {
+        let (cxx, rust) =
+            check_legacy_sigop_count_script::<CxxInterpreter, RustInterpreter>(script);
+        if rust != cxx {
+            warn!(
+                "The Rust Zcash Script interpreter had a different sigop count ({:?}) from the C++ one ({:?}).",
+                rust,
+                cxx)
+        };
+        cxx
+    }
+
+    fn verify_callback(
+        sighash: SighashCalculator,
+        lock_time: i64,
+        is_final: bool,
+        script_pub_key: &[u8],
+        script_sig: &[u8],
+        flags: VerificationFlags,
+    ) -> Result<(), Error> {
+        let (cxx, rust) = check_verify_callback::<CxxInterpreter, RustInterpreter>(
+            sighash,
+            lock_time,
+            is_final,
+            script_pub_key,
+            script_sig,
+            flags,
+        );
+        if rust.map_err(normalize_error) != cxx {
+            // probably want to distinguish between
+            // - C++ succeeding when Rust fails (bad),
+            // - Rust succeeding when C++ fals (worse), and
```

Comment:
typo

Reply:
Also, both are really bad because they can be used to force a consensus fork, during the time when both implementations are in use.

Reply:
🤔